### PR TITLE
New timer conditions with an operator.

### PR DESCRIPTION
**Pull request by D8H**
Implementation for:
* https://github.com/4ian/GDevelop/discussions/3446
* https://github.com/4ian/GDevelop/issues/3282

It still allows every operator.

Example with scene and object timers:
* https://www.dropbox.com/s/zmd89fjpozt7cu2/Permissive.zip?dl=1

### CODE REVIEW

File: Core/GDCore/Extensions/Builtin/BaseObjectExtension.cpp [hunk 1]
```diff
@@ -856,7 +857,21 @@ void GD_CORE_API BuiltinExtensionsImplementer::ImplementsBaseObjectExtension(
          "res/conditions/timer.png")
       .AddParameter("object", _("Object"))
       .AddParameter("string", _("Timer's name"))
-      .AddParameter("expression", _("Time in seconds"));
+      .AddParameter("expression", _("Time in seconds"))
+      .SetHidden();
+
+  obj.AddCondition("ObjectTimerComparison",
```
Comment (4ian) [inline Core/GDCore/Extensions/Builtin/BaseObjectExtension.cpp:863]: Let's use a verb for the name: `CompareObjectTimer`

File: Core/GDCore/Extensions/Builtin/TimeExtension.cpp [hunk 2]
```diff
@@ -33,7 +33,22 @@ void GD_CORE_API BuiltinExtensionsImplementer::ImplementsTimeExtension(
                     "res/conditions/timer.png")
       .AddCodeOnlyParameter("currentScene", "")
       .AddParameter("expression", _("Time in seconds"))
-      .AddParameter("string", _("Timer's name"));
+      .AddParameter("string", _("Timer's name"))
+      .SetHidden();
+
+  extension
+      .AddCondition("TimerComparison",
```
Comment (4ian) [inline Core/GDCore/Extensions/Builtin/TimeExtension.cpp:40]: And let's use `CompareTimer` here :)

File: GDJS/Runtime/events-tools/runtimescenetools.ts [hunk 3]
```diff
@@ -106,11 +123,35 @@ namespace gdjs {
         return timeManager.getTimer(timerName).getTime() / 1000;
       };
 
-      export const getTimeFromStartInSeconds = function (runtimeScene) {
+      /**
+       * This is used by the condition to return false when a timer doesn't exist
+       * no mater the relational operator.
```
Comment (4ian) [inline GDJS/Runtime/events-tools/runtimescenetools.ts:128]:
```suggestion
       * no matter the relational operator.
```

File: GDJS/Runtime/events-tools/runtimescenetools.ts [hunk 4]
```diff
@@ -106,11 +123,35 @@ namespace gdjs {
...
+       */
+      export const getTimerElapsedTimeInSecondsOrNaN = function (
+        runtimeScene: gdjs.RuntimeScene,
+        timerName: string
+      ) {
+        const timeManager = runtimeScene.getTimeManager();
+        if (!timeManager.hasTimer(timerName)) {
+          return Number.NaN;
+        }
+        console.log(timeManager.getTimer(timerName).getTime() / 1000);
```
Comment (4ian) [inline GDJS/Runtime/events-tools/runtimescenetools.ts:141]: Let's remove the console.log ;)

File: GDJS/Runtime/runtimeobject.ts [hunk 5]
```diff
@@ -1780,6 +1787,22 @@ namespace gdjs {
       return this._timers.get(timerName).getTime() / 1000.0;
     }
 
+    /**
+     * Get a timer elapsed time.
+     *
+     * This is used by the condition to return false when a timer doesn't exist
+     * no mater the relational operator.
```
Comment (4ian) [inline GDJS/Runtime/runtimeobject.ts:1794]: ditto here

**Review by 4ian — COMMENTED** [review]
I know this is a draft but still added a review ;)

**Comment by 4ian** [comment]
Looks great, thank you! 
@charburgx Good to keep in mind for next tutorials => the usage of timer conditions now include the operator :) 

@D8H @ClementPasteau @AlexandreSi  I wonder if for conditions like this, we could in the editor select a default operator (greater than)?

**Comment by ClementPasteau** [comment]
Yeah that could be a nice touch of UX, not having to select the operator, defaulted to the most used.
And we could do that for most conditions/actions too. It's a tiny time saver, but can add up to quite a lot :)

**Comment by D8H** [comment]
Yes, the parameter could be marked as optional with a default value.

What do you think of some sort of radio buttons instead of a combobox?
Classic radio buttons take a lot more space and it's harder to separate visually. I wonder if the symbol could be the radio button itself. It would allow to have a row of small buttons (like a switch but with more than 2 states), but we loose the labels (though the symbols might be enough and they probably there only to fill the combobox).

**Comment by D8H** [comment]
It's not very pretty, but it's not ugly either. It could make it faster to choose an operation while we think of a better look and feel.

operatorButtons

**Comment by 4ian** [comment]
As you say, it's not the prettiest thing nor super ugly. I wonder if we could group them?
https://v4.mui.com/components/button-group/

And as order: <, <=, >=, > ?

**Comment by D8H** [comment]
The group button seems to be for having the same look on each button (it seems to override the syle of its children)

I removed the margin and used FlatButton for the unselected ones. I think it has a modern look. Maybe a background color could give more cohesiveness, but I kind of like it like this.

image

**Comment by tristanbob** [comment]
I like the idea of offering these operators, but it doesn't feel intuitive yet.  It is not obvious that there are 3 parameters that can be changed.

How about a drop-down list that showed the operator and the text description?

```
(field) Timer's name 

(pulldown) 
< "Is lesser than"
<= "Is lesser than or equal" 
>= "Is greater than or equal"  <default selection>
> "Is greater than"

(field) Time in seconds 
```

**Comment by tristanbob** [comment]
I guess I should have just said, I like the consistency of doing the same comparison we do for Variables ("=" and "!=" can be removed)

image

**Comment by D8H** [comment]
> I guess I should have just said, I like the consistency of doing the same comparison we do for Variables ("=" and "!=" can be removed)

If we change the operator component, it will change everywhere.

**Pull request merged**